File: app/src/main/java/com/example/islamic/view/HadithListActivity.java
package com.example.islamic.view;

import android.content.Intent;
import android.os.Bundle;
import android.widget.Toast;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import com.example.islamic.R;
import com.example.islamic.adapter.HadithAdapter;
import com.example.islamic.model.Hadith;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;
import java.io.IOException;
import java.io.InputStream;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;

public class HadithListActivity extends AppCompatActivity {

    private RecyclerView recyclerView;
    private ArrayList<Hadith> hadithList = new ArrayList<>();
    private HadithAdapter adapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_hadith_list);

        recyclerView = findViewById(R.id.hadith_list_rv);
        recyclerView.setLayoutManager(new LinearLayoutManager(this));

        adapter = new HadithAdapter(hadithList, hadith -> {
            Intent intent = new Intent(HadithListActivity.this, HadithDetailActivity.class);
            intent.putExtra("hadith_object", hadith);
            startActivity(intent);
        });

        recyclerView.setAdapter(adapter);
        loadHadithData();
    }

    private void loadHadithData() {
        try {
            InputStream is = getAssets().open("hadith.json");
            int size = is.available();
            byte[] buffer = new byte[size];
            is.read(buffer);
            is.close();
            String json = new String(buffer, StandardCharsets.UTF_8);
            JSONArray jsonArray = new JSONArray(json);

            hadithList.clear();

            for (int i = 0; i < jsonArray.length(); i++) {
                JSONObject obj = jsonArray.getJSONObject(i);
                String fullHadith = obj.getString("hadith");
                String description = obj.getString("description");

                // استخراج العنوان ونص الحديث من حقل "hadith"
                String[] parts = fullHadith.split("\n\n", 2);
                String title = parts.length > 0 ? parts[0] : "حديث " + (i + 1);
                String hadithText = parts.length > 1 ? parts[1] : fullHadith;

                hadithList.add(new Hadith(title, hadithText, description));
            }

            adapter.notifyDataSetChanged();

        } catch (IOException | JSONException e) {
            e.printStackTrace();
            Toast.makeText(this, "خطأ في تحميل ملف الأحاديث!", Toast.LENGTH_LONG).show();
        }
    }
}

File: app/src/main/java/com/eslamdev/islamic/presentation/ui/activity/HadithListActivity.kt
package com.eslamdev.islamic.presentation.ui.activity

import android.content.Intent
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.LinearLayoutManager
import com.eslamdev.islamic.data.repository.HadithRepository
import com.eslamdev.islamic.databinding.ActivityHadithListBinding
import com.eslamdev.islamic.presentation.ui.adapter.HadithAdapter
import com.eslamdev.islamic.presentation.viewmodel.HadithViewModel
import com.eslamdev.islamic.presentation.viewmodel.HadithViewModelFactory
import kotlinx.coroutines.flow.collectLatest

class HadithListActivity : AppCompatActivity() {

    private lateinit var binding: ActivityHadithListBinding
    private lateinit var viewModel: HadithViewModel
    private val adapter = HadithAdapter { hadith ->
        val intent = Intent(this, HadithDetailActivity::class.java)
        intent.putExtra("hadith_object", hadith)
        startActivity(intent)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityHadithListBinding.inflate(layoutInflater)
        setContentView(binding.root)

        val repository = HadithRepository(this)
        val factory = HadithViewModelFactory(repository)
        viewModel = ViewModelProvider(this, factory)[HadithViewModel::class.java]
        binding.btnBack.setOnClickListener{
            finish()
        }
        setupRecyclerView()

        lifecycleScope.launchWhenStarted {
            viewModel.hadiths.collectLatest { list ->
                adapter.submitList(list)
            }
        }

        viewModel.loadHadiths()
    }

    private fun setupRecyclerView() {
        binding.hadithListRv.layoutManager = LinearLayoutManager(this)
        binding.hadithListRv.adapter = adapter
    }
}